User `uv pip compile` to recompile requirements files

toofar opened the pull request:
<!-- Thanks for submitting a pull request! Please pick a descriptive title (not just "issue 12345"). If there is an open issue associated to your PR, please add a line like "Closes #12345" somewhere in the PR description (outside of this comment) -->

The-Compiler reviewed (changes requested):
Looking great, I left a couple of nitpicks. Maybe it'd be good to leave out the autogenerated changes from the PR, so we avoid conflicts and I can just update them after merging?

Can't wait to start using `uv` after trying it out recently. Probably will take a quick look at installing [tox-uv](https://github.com/tox-dev/tox-uv) using tox' [self-bootstrapping setting](https://tox.wiki/en/4.23.2/config.html#requires) as well. Probably the testing part of the recompile script will be much faster with that enabled too!

toofar commented on scripts/dev/recompile_requirements.py:
```diff
@@ -24,18 +24,47 @@
+    return "==".join([normalize_pkg(pkg), version])
+
+
+CHANGELOG_URLS = {normalize_pkg(name): url for name, url in CHANGELOG_URLS.items()}
```
In general I would prefer removing points of friction for contributing, and maintaining. In this case I think it would be nice if people could put packages names in the json files in whatever case they like. The robots doing comparisons should be the ones that have to care about normalization. For example, `PyQt5` and `PyJWT` are in there at the moment, nothing wrong with writing them like that.

toofar commented on scripts/dev/recompile_requirements.py:
```diff
@@ -346,31 +396,71 @@ def get_outfile(name):
 
-def build_requirements(name):
-    """Build a requirements file."""
+def get_updated_requirements_freeze(filename, venv_dir, comments):
```
If you are happy to get rid of it, then I'm happy to get rid of it! I just left it in there to allow for easy comparisons easy switching back if we aren't happy with the pip compile based implementation.

toofar commented on scripts/dev/recompile_requirements.py:
```diff
@@ -346,31 +396,71 @@ def get_outfile(name):
+              pre=comments['pre'],
+              pip_args=comments['pip_args'])
+    with utils.gha_group('Freezing requirements'):
+        args = ['--all'] if '-tox.txt-raw' in filename else []
```
Sure, I can change that. But I would like to hear your opinion on removing the special casing of setuptools and pip entirely, which would remove this block altogether. `pip freeze` excludes it's own dependancies from it's output to avoid polluting peoples requirements lists, because its starting point is looking at all installed packages, and then we have to add these excluded requirements back when we actually want them. But compile based approaches like pip-compile that have a starting point of a requirements file don't have that problem.

See upstream discussions around including setuptools that I linked in this commit: https://github.com/qutebrowser/qutebrowser/pull/8276/commits/a4ca2f9cac2319bc0c260ab63368f14cbe09e9d2

toofar commented:
Thanks for the review! I've addressed most of your comments. I've:

1. rebased, so this is up to date with the latest requirements files
2. dropped almost all the automated changes. Still did a manual run at the end though else the new format of the requirement files will come through next time the update job runs, which might look a bit alarming. IMO this makes it harder to review the PR commit by commit and build up trust of the changes because you can't see the output alongside the commits that caused them. I can add back incremental autogenerated changes if that becomes an issue.
3. removed the old `pip freeze` based method
4. removed the code filtering pip and setuptools from the compiled requirements files, happy to add that back if you want
